Implement BPF-based masquerading for IPv6

pchaigno opened the pull request:
Summary of commits:
1. Preparatory: use a bit of inline BPF asm to avoid bad optimization from LLVM.
2. Preparatory: refactor `ct_is_reply4` as a macro to facilitate implem. of IPv6 counterpart.
3. Implement the datapath changes by following the IPv4 implementation.
4. Preparatory: refactor `InitBPFMasqueradeAddrs` to facilitate implem. of IPv6 counterpart.
5. Implement agent changes for the flags and macros.
6. Update documentation.
7. Extend the tests to cover IPv6 BPF masquerading.

See commits for details.

Fixes: https://github.com/cilium/cilium/issues/14350.

pchaigno committed:
daemon, config: Enable IPv6 BPF masquerading

This commit allows enabling BPF masquerading for IPv6 and defines all
the proper macros for the datapath implementation.

[ Quentin: Rebased on master branch, addressed minor conflicts on
  daemon.go and loader.go. ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
bpf: Implement IPv6 BPF masquerading

This commit implements BPF masquerading for IPv6 by following the
implementation for IPv4. The main function changed here is
snat_v6_needed with almost the exact same steps as IPv4.

Support for the ipmasq agent in IPv6 is not implemented here.

[ Quentin: Rebased on master branch, addressed minor conflicts in
  nodeport.h due to commit 0ea434115253 ("bpf: Make union v6addr const
  where possible"). ]

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno committed:
bpf: Avoid 32bit assignment of packet pointer

Since ctx->{data,data_end,data_meta} are 32-bits fields, LLVM sometimes
generates 32-bit assignments for those pointers. This leads to verifier
errors as the verifier is unable to track the packet pointers through
the 32-bit assignments, as show below.

    ; return (void *)(unsigned long)ctx->data;
    2: (61) r9 = *(u32 *)(r7 +76)
    ; R7_w=ctx(id=0,off=0,imm=0) R9_w=pkt(id=0,off=0,r=0,imm=0)
    ; return (void *)(unsigned long)ctx->data;
    3: (bc) w6 = w9
    ; R6_w=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R9_w=pkt(id=0,off=0,r=0,imm=0)
    ; if (data + tot_len > data_end)
    4: (bf) r2 = r6
    ; R2_w=inv(id=1,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6_w=inv(id=1,umax_value=4294967295,var_off=(0x0; 0xffffffff))
    5: (07) r2 += 54
    ; R2_w=inv(id=0,umin_value=54,umax_value=4294967349,var_off=(0x0; 0x1ffffffff))
    ; if (data + tot_len > data_end)
    6: (2d) if r2 > r1 goto pc+466
    ; R1_w=pkt_end(id=0,off=0,imm=0) R2_w=inv(id=0,umin_value=54,umax_value=4294967349,var_off=(0x0; 0x1ffffffff))
    ; tmp = a->d1 - b->d1;
    7: (71) r2 = *(u8 *)(r6 +22)
    R6 invalid mem access 'inv'

As a workaround, Yonghong Song suggested to read those fields using asm
bytecode. With that trick, LLVM is unaware that the original field is
actually on 32 bits and can't perform the above "optimization".

This workaround is needed only now because a subsequent commit
introduces some BPF C code that causes this bytecode to be generated
(snat_v6_needed).

Related: https://lore.kernel.org/bpf/c5a76b4d-abed-51f6-bf16-040eb0baf290@fb.com/T/#m933626f7404562a6e98af78a8f44c30977681ffa
Suggested-by: Yonghong Song <yhs@fb.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Test BPF masquerading for IPv6

Since Cilium supports BPF masquerading for IPv6, we can now test it.

The upgrade/downgrade test needs a small exception because IPv6 BPF
masquerading is still not supported on v1.11, from which we upgrade.
The exception logic comes with an assertion, to ensure we don't forget
to remove the exception once the upgrade is performed from v1.12.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Refactor ct_is_reply as a macro

This will allow us to easily define the IPv6 equivalent in a subsequent
commit. This commit should include no functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
node: Refactor InitBPFMasqueradeAddrs

This commit refactors InitBPFMasqueradeAddrs to easily extend it for
IPv6 in a subsequent commit. This commit shouldn't have any functional
changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

christarazi reviewed (approved):
:rocket:

qmonnet reviewed (approved):
Looks good, nice work!

Maybe one more thing to update:
```
$ grep supported Documentation/concepts/networking/masquerading.rst
    eBPF based masquerading is currently not supported for IPv6 traffic.
```

pchaigno committed:
docs: Update documentation for BPF IPv6 masquerading

BPF-based IPv6 masquerading is now supported, except for the ip-masq
agent. This commit therefore updates the documentation accordingly.

Signed-off-by: Paul Chaignon <paul@cilium.io>

qmonnet reviewed (approved):
> I updated the documentation in the commit before last.

Commit before last looks good!

brb commented on bpf/lib/conntrack.h:
```diff
@@ -267,6 +267,33 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
+ */
+#define CT_IS_REPLY(FAMILY)						\
+static __always_inline int						\
+ct_is_reply ## FAMILY(const void *map, struct __ctx_buff *ctx, int off,	\
```
Ha, neat! I think more functions in the conntrack.h might benefit from this trick.

Maybe s/CT_IS_REPLY/DEFINE_FUNC_CT_IS_REPLY/ to make it more clear what the macro does.

pchaigno replied:
To be fair, I didn't invent this trick. We're already doing that a lot for the proxy and ipcache logic.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -194,39 +198,125 @@ static __always_inline bool nodeport_uses_dsr6(const struct ipv6_ct_tuple *tuple
 		return false;
-#ifdef ENABLE_DSR_HYBRID
+
+	/* We need to SNAT when there is a potential of overlapping tuples, e.g.
```
Nit: not sure what is a convention here, but I usually try to avoid duplicating v4/v6 comments, and instead I comment only the v4 equivalent. Otherwise, we are risking that comments might get stale.

brb commented on bpf/lib/conntrack.h:
```diff
@@ -606,30 +633,7 @@ ct_extract_ports4(struct __ctx_buff *ctx, int off, enum ct_dir dir,
-
-	return 0;
-}
+CT_IS_REPLY(4)
```
Maybe add a comment `/* This defines ct_is_reply() function */`, so it can be grep-able?

brb reviewed (approved):
Nice! Small nits.

ldelossa commented on bpf/include/bpf/ctx/skb.h:
```diff
@@ -57,6 +57,15 @@
+{
+	void *data;
+
+	asm volatile("%0 = *(u32 *)(%1 + %2)" : "=r"(data) : "r"(ctx), "i"(76));
```
Can we throw a comment or some indication to why we use inline byte code here, especially when we see its not used on XDP side. I think it could throw off the reader unless they git blame.

pchaigno replied:
Turns out it's needed for XDP as well. I missed it before because the tests were disabled :grimacing:

ldelossa reviewed (approved):
looks awesome, just a small nit about comments around inline byte code.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -181,39 +181,97 @@ static __always_inline bool nodeport_uses_dsr6(const struct ipv6_ct_tuple *tuple
 static __always_inline bool snat_v6_needed(struct __ctx_buff *ctx,
```
```
/* TODO(brb): after GH#6320, we can move snat_v{4,6}_needed() to lib/nat.h, as
 * then the helper function won't depend the dsr checks.
 */
```

Can we resolve this TODO now as well?

pchaigno replied:
Please do.
